math/complex_numbers.cpp

```diff
@@ -0,0 +1,238 @@
+     * @return result current number subtract other number
+     */
+    Complex operator-(const Complex &other) {
+        Complex result(this->re - other.re, this->im - other.im);
+        return result;
+    }
+
+    /**
+     * Operator overload to be able to multiple two complex numbers.
+     * @param other The other number to multiply the current number to.
+     * @return result current number times other number.
+     */
+    Complex operator*(const Complex &other) {
+        Complex result(this->re * other.re - this->im * other.im,
+                       this->re * other.im + this->im * other.re);
+        return result;
+    }
+
+    /**
+     * Operator overload of the BITWISE NOT which gives us the conjugate of our
+     * complex number. NOTE: This is overloading the BITWISE operator but its
+     * not a BITWISE operation in this definition.
+     * @return result The conjugate of our complex number.
+     */
+    Complex operator~() const {
+        Complex result(this->re, -(this->im));
+        return result;
+    }
+
+    /**
+     * Operator overload to be able to divide two complex numbers. This function
+     * would throw an exception if the other number is zero.
+     * @param other The other number we divide our number by.
+     * @return result Current number divided by other number.
+     */
+    Complex operator/(const Complex &other) {
+        Complex result = *this * ~other;
+        double denominator =
+            other.real() * other.real() + other.imag() * other.imag();
+        if (denominator != 0) {
+            result = Complex(result.real() / denominator,
+                             result.imag() / denominator);
+            return result;
+        } else {
+            throw std::invalid_argument("Undefined Value");
+        }
+    }
+
+    /**
+     * Operator overload to be able to copy RHS instance of Complex to LHS
+     * instance of Complex
+     */
+    const Complex &operator=(const Complex &other) {
+        this->re = other.real();
+        this->im = other.imag();
+        return *this;
+    }
+};
+
+/**
+ * Logical Equal overload for our Complex class.
+ * @param a Left hand side of our expression
+ * @param b Right hand side of our expression
+ * @return 'True' If real and imaginary parts of a and b are same
+ * @return 'False' Otherwise.
+ */
+bool operator==(const Complex &a, const Complex &b) {
+    return a.real() == b.real() && a.imag() == b.imag();
+}
+
+/**
+ * Overloaded insersion operator to accommodate the printing of our complex
+ * number in their standard form.
+ * @param os The console stream
+ * @param num The complex number.
+ */
+std::ostream &operator<<(std::ostream &os, const Complex &num) {
+    os << "(" << num.real();
+    if (num.imag() < 0) {
+        os << " - " << -num.imag();
+    } else {
+        os << " + " << num.imag();
+    }
+    os << "i)";
+    return os;
+}
+
+/**
+ * Tests Function
+ */
+void tests() {
+    Complex num1(1, 1), num2(1, 1);
```

Comment:
would it be better to replace the hard-coded numbers by something like:
```suggestion
    double get_rand(){
       return (std::rand() % 100 - 50) / 100.f;
    }

    std::srand(std::time(nullptr));
    Complex num1(get_rand(), get_rand()), num2(get_rand(), get_rand());
```
What are the pros and cons?

Reply:
None that I can foresee.